fab/utils/numerical.py

```diff
@@ -22,12 +22,22 @@ def effective_sample_size(log_w: torch.Tensor, normalised=False):
         log_w = F.softmax(log_w, dim=0)
     return 1 / torch.sum(log_w ** 2) / log_w.shape[0]
 
+def effective_sample_size_over_p(log_w: torch.Tensor):
+    """Estimate the effective sample size using samples from the target.
+    Note that these log weights must be calculated with a normalised target log prob."""
+    # effective sample size, see https://arxiv.org/abs/1602.03572
+    assert len(log_w.shape) == 1
+    return 1 / torch.mean(torch.exp(log_w))
+
+
 
-def quadratic_function(x: torch.Tensor):
+def quadratic_function(x: torch.Tensor, seed: int = 0):
+    torch.random.manual_seed(seed)
     # example function that we may want to calculate expectations over
     x_shift = 2*torch.randn(x.shape[-1]).to(x.device)
     A = 2*torch.rand((x.shape[-1], x.shape[-1])).to(x.device)
     b = torch.rand(x.shape[-1]).to(x.device)
+    torch.seed() # set back to random number
```

Comment:
Have to make sure that the randomness for the quadratic function is dealt with correctly so that it is the same for each evaluation.